File: 2-add-two-numbers/2-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
         if(l1 == null || l2 == null)
        {
            return l1 == null ? l2 : l1; 
        }
        
        ListNode dummy = new ListNode(-1);
        ListNode prev = dummy; 
        int carry = 0; 
        ListNode p1 = l1, p2 = l2; 
        while(p1 != null || p2 != null || carry != 0)
        {
            int sum = 0;
            if(p1 != null)
            {
                sum += p1.val;
                p1 = p1.next; 
            }

            if(p2 != null)
            {
                sum += p2.val;
                p2 = p2.next;  
            }

        sum += carry; 

        carry = sum/10; 

        ListNode node = new ListNode(sum%10);
        prev.next = node;
        prev = prev.next;
        }

        return dummy.next; 
    }
}

File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        
        if(l1 == nullptr || l2 == nullptr)
        {
            return l1 == nullptr ? l2 : l1; 
        }
        
        ListNode* dummy = new ListNode(); 
        ListNode* prev = dummy; 
        

        int carry = 0; 
        
        while(l1 != nullptr || l2 != nullptr || carry)
        {
            
            int sum = 0; 
            
            if(l1 != nullptr)
            {
                sum += l1->val;
                l1 = l1->next; 
            }
            
            
            if(l2 != nullptr)
            {
                sum += l2->val; 
                l2 = l2->next; 
            }
            
            sum += carry; 
            
            carry = sum/10; 
            
            ListNode* node = new ListNode(sum%10); 
            
            prev->next = node; 
            
            prev = prev->next; 
            
        }
        
        return dummy->next;         
    
    }
};